### PR TITLE
Fix suppression matching to include files

### LINKED ISSUE #3983: SuppressedIssuesMatcher matches issues in wrong files
### Description

[SuppressedIssueMatcher ](https://github.com/SonarSource/sonarlint-visualstudio/blob/master/src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs#L66)does not take into account the file path. When suppressions change (via event or timer), all suppressions that have the identical line are suppressed, irrespective of the file the user suppressed them in.

Repro:
1. Create file a.cs with a violation
```
public class A { 
// todo: do something 
} 
```
2. Create b.cs with the same exact violation
```
public class B { 
// todo: do something 
} 
```
3. Suppress issue of file A on the server

Expected ==> only A is suppressed
Actual ==> both A and B are suppressed

Note:
* As long as the line hash is identical and it's the same rule id, all instances of this violation will be suppressed (and un-suppressed)

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #3983

### CODE REVIEW

File: src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs [hunk 1]
```diff
@@ -70,6 +73,11 @@ private static bool IsMatch(IFilterableIssue issue, SonarQubeIssue serverIssue)
                 return false;
             }
 
+            if (!IsSameFile(issue, serverIssue))
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:76]: Strictly speaking the method is also checking whether the issues are both module-level (i.e. don't have a file). I couldn't think of a better name, though.

File: src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs [hunk 2]
```diff
@@ -78,5 +85,36 @@ private static bool IsMatch(IFilterableIssue issue, SonarQubeIssue serverIssue)
...
+            Debug.Assert(!Path.IsPathRooted(serverPath) && !serverPath.Contains("/"),
+                $"Expecting the server-side file path to be relative and not to contain forward-slashes.");
+
+            Debug.Assert(!serverPath.StartsWith("\\"), "Not expecting server file path to start with a back-slash");
+            if(localPath.EndsWith(serverPath, StringComparison.OrdinalIgnoreCase))
+            {
+                // Check the preceding local character is a backslash  - we want to make sure a server path
+                // of `aaa\foo.txt` matches `c:\aaa\foo.txt` but not `c:`bbbaaa\foo.txt`
+                return localPath.Length > serverPath.Length &&
+                    localPath[localPath.Length - serverPath.Length - 1] == '\\';
```
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: @rita-sonarsource thanks. It looks like both of the input paths are expected to be full paths, but I'll check that method and other ones in `PathHelper`.

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: @rita-sonarsource there are a number of places where we try to match a relative server path to a local absolute path
1) here
2) in the equivalent Roslyn suppression handling code (`SuppressionChecker`)
3) for hotspots and taint (when working out which local file a server issue maps to).

(3) is ok; it uses `IAbsoluteFilePathLocator`, which internally uses a VS API to go from a relative path to an absolute (i.e. does this relative path map to any absolute paths).

(2) isn't ok - server path `aaa\foo.txt` will match local path `c:`bbbaaa\foo.txt`

I'll commit this as-is, create a separate issue for (2) and look at making the (1) and (2) use the same code.

**Review by bigfluffycookie — APPROVED** [review]
Looks good, as duncan said I think it makes sense to create a separate issue for the problem in the comments.

File: src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: @duncanp-sonar how did you use ResolveRelativePath? with the solution directory?
I thought that new Uri will add automatically "/", so `bbbaaa` will never match, as it should be `bb\...`

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: @rita-sonarsource `ResolveRelativePath` combines a relative path and an absolute path and returns the result. It doesn't tell us if the relative part is contained in the absolute part or not.

e.g. `PathHelper.ResolveRelativePath(x, y) =>`

```
( aaa\file1.ext, x:\aaa\file1.txt ) => x:\aaa\aaa\file1.ext

( aaa\file.ext", x:\xxx\aaa\file1.txt ) => x:\xxx\aaa\aaa\file1.ext

( aaa\file1.ext", x:\bbbaaa\file1.txt ) => x:\bbbaaa\aaa\file1.ext

( aaa\file2.ext, x:\xxx\aaa\file1.txt ) => x:\xxx\aaa\aaa\file2.ext

```

[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: @duncanp-sonar I understand, so we can then use it to compare.

```
if (localPath.Equals(pathHelper.ResolveRelativePath(serverPath, solutionFolder))
```
WDYT?

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: Sorry, I missed that you'd added in the solution folder.

We don't know what the solution folder is at this point, and the file paths don't have to be under it anyway.

[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:114]: We can get it from the IWorkspaceService, and we know that everything has to be under it when comparing.

File: src/ConnectedMode.UnitTests/Suppressions/SuppressedIssueMatcherTests.cs [hunk 3]
```diff
@@ -129,18 +129,53 @@ public void MatchExists_ResultDependsOnSuppressionState(string serverRuleId, int
...
+        [DataRow("", @"c:\any.txt", false)]
+        [DataRow(null, @"c:\any.txt", false)]
+        
+        // File issues
+        [DataRow(@"same.txt", @"c:\same.txt", true)]
+        [DataRow(@"SAME.TXT", @"c:\same.txt", true)]
+        [DataRow(@"same.TXT", @"c:\XXXsame.txt", false)]  // partial file name -> should not match
+        [DataRow(@"differentExt.123", @"a:\differentExt.999", false)] // different extension -> should not match
+        [DataRow(@"aaa\partial\file.cs", @"d:\partial\file.cs", false)]
+        // Only matching the local path tail, so the same server path can match multiple local files
```
Comment (rita-sonarsource) [inline src/ConnectedMode.UnitTests/Suppressions/SuppressedIssueMatcherTests.cs:151]: @duncanp-sonar the last 3 unit test cases here are what I mean by incorrect behavior. And I don't see how you can fix it without getting the solution folder.

Comment (rita-sonarsource) [reply src/ConnectedMode.UnitTests/Suppressions/SuppressedIssueMatcherTests.cs:151]: Edit: as we discussed offline, this was the case in the old code as well, and what the other SL flavors are doing as well. So let's stay with that.

**Review by rita-sonarsource — APPROVED** [review]

**Pull request merged**